File: web/src/restore.ts
import { existsSync } from "node:fs";
import { readFile, writeFile } from "node:fs/promises";
import { join } from "node:path";

import { status } from "./status.js";
import {
  extractContentFromBlob,
  extractTreeFromCommit,
  parseTreeEntries,
  readObject,
} from "./utils.ts";

export async function restore(path: string, files: string[]): Promise<void> {
  const gitDir = join(path, ".git");

  if (!existsSync(gitDir)) {
    throw new Error("Not a git repository");
  }

  for (const file of files) {
    const filePath = join(path, file);

    if (!existsSync(filePath)) {
      throw new Error(`File not found: ${file}`);
    }
  }

  const branchPath = join(gitDir, "refs", "heads", "main");
  const commitSha = (await readFile(branchPath, "utf-8")).trim();

  const commitData = await readObject(gitDir, commitSha);
  const treeSha = extractTreeFromCommit(commitData);

  for (const file of files) {
    const blobSha = await findBlobInTree(gitDir, treeSha, file);
    if (blobSha === null) {
      throw new Error(`File not in commit: ${file}`);
    }

    const blobData = await readObject(gitDir, blobSha);
    const content = extractContentFromBlob(blobData);
    const filePath = join(path, file);
    await writeFile(filePath, content, "utf-8");
  }
}

export async function restoreAll(path: string): Promise<void> {
  const gitDir = join(path, ".git");

  if (!existsSync(gitDir)) {
    throw new Error("Not a git repository");
  }

  const currentStatus = await status(path);
  const filesToRestore = [...currentStatus.modified];

  if (filesToRestore.length === 0) {
    return;
  }

  await restore(path, filesToRestore);
}

async function findBlobInTree(
  gitDir: string,
  treeSha: string,
  filePath: string,
): Promise<string | null> {
  const parts = filePath.split("/");
  const [name, ...rest] = parts;

  const treeData = await readObject(gitDir, treeSha);
  const entries = parseTreeEntries(treeData);

  for (const entry of entries) {
    if (entry.path === name) {
      if (entry.type === "blob") {
        if (rest.length === 0) {
          return entry.sha;
        }
        return null;
      }
      if (entry.type === "tree") {
        if (rest.length > 0) {
          return findBlobInTree(gitDir, entry.sha, rest.join("/"));
        }
        return null;
      }
    }
  }

  return null;
}

File: dotnet/Gitologist/Restore.cs
using Gitologist.Types;

namespace Gitologist;

public static class Restore
{
    public static async Task RestoreFiles(string path, string[] files)
    {
        var gitDir = Path.Combine(path, ".git");

        if (!Directory.Exists(gitDir))
        {
            throw new InvalidOperationException("Not a git repository");
        }

        foreach (var file in files)
        {
            var filePath = Path.Combine(path, file);

            if (!File.Exists(filePath))
            {
                throw new FileNotFoundException($"File not found: {file}");
            }
        }

        var branchPath = Path.Combine(gitDir, "refs", "heads", "main");
        var commitSha = (await File.ReadAllTextAsync(branchPath)).Trim();

        var commitData = await Utils.ReadObject(gitDir, commitSha);
        var treeSha = Utils.ExtractTreeFromCommit(commitData);

        foreach (var file in files)
        {
            var blobSha = await FindBlobInTree(gitDir, treeSha, file);
            if (blobSha == null)
            {
                throw new InvalidOperationException($"File not in commit: {file}");
            }

            var blobData = await Utils.ReadObject(gitDir, blobSha);
            var content = Utils.ExtractContentFromBlob(blobData);
            var filePath = Path.Combine(path, file);
            await File.WriteAllTextAsync(filePath, content);
        }
    }

    public static async Task RestoreAll(string path)
    {
        var gitDir = Path.Combine(path, ".git");

        if (!Directory.Exists(gitDir))
        {
            throw new InvalidOperationException("Not a git repository");
        }

        var currentStatus = await Status.GetStatus(path);
        var filesToRestore = currentStatus.Modified.ToList();

        if (filesToRestore.Count == 0)
        {
            return;
        }

        await RestoreFiles(path, filesToRestore.ToArray());
    }

    private static async Task<string?> FindBlobInTree(
        string gitDir,
        string treeSha,
        string filePath
    )
    {
        var parts = filePath.Split('/');
        var name = parts[0];
        var rest = parts.Skip(1).ToArray();

        var treeData = await Utils.ReadObject(gitDir, treeSha);
        var entries = Utils.ParseTreeEntries(treeData);

        foreach (var entry in entries)
        {
            if (entry.Path == name)
            {
                if (entry.Type == "blob")
                {
                    if (rest.Length == 0)
                    {
                        return entry.Sha;
                    }

                    return null;
                }

                if (entry.Type == "tree")
                {
                    if (rest.Length > 0)
                    {
                        return await FindBlobInTree(
                                gitDir,
                                entry.Sha,
                                string.Join("/", rest)
                            );
                    }

                    return null;
                }
            }
        }

        return null;
    }
}
